File: src/main/java/org/tobi29/scapes/engine/gui/GuiControllerMouse.java
package org.tobi29.scapes.engine.gui;

import java8.util.Optional;
import java8.util.stream.Stream;
import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.input.ControllerBasic;
import org.tobi29.scapes.engine.input.ControllerDefault;
import org.tobi29.scapes.engine.input.ControllerKey;
import org.tobi29.scapes.engine.utils.Pair;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.math.vector.Vector2d;

public class GuiControllerMouse extends GuiControllerDefault {
    private final double scrollSensitivity;
    private final GuiCursor cursor = new GuiCursor(false);
    private Optional<GuiComponent> draggingLeft = Optional.empty(),
            draggingRight = Optional.empty();
    private double dragLeftX, dragLeftY, dragRightX, dragRightY;

    public GuiControllerMouse(ScapesEngine engine,
            ControllerDefault controller) {
        this(engine, controller, 1.0);
    }

    public GuiControllerMouse(ScapesEngine engine, ControllerDefault controller,
            double scrollSensitivity) {
        super(engine, controller);
        this.scrollSensitivity = scrollSensitivity;
    }

    @Override
    public void update(double delta) {
        double cursorX = controller.x();
        double cursorY = controller.y();
        double ratio = 540.0 / engine.container().containerHeight();
        double guiCursorX = cursorX * ratio;
        double guiCursorY = cursorY * ratio;
        cursor.set(new Vector2d(cursorX, cursorY),
                new Vector2d(guiCursorX, guiCursorY));
        if (draggingLeft.isPresent()) {
            GuiComponent component = draggingLeft.get();
            double relativeX = cursor.guiX() - dragLeftX;
            double relativeY = cursor.guiY() - dragLeftY;
            dragLeftX = cursor.guiX();
            dragLeftY = cursor.guiY();
            component.gui().sendNewEvent(GuiEvent.DRAG_LEFT,
                    new GuiComponentEvent(guiCursorX, guiCursorY, relativeX,
                            relativeY), component, engine);
        }
        if (draggingRight.isPresent()) {
            GuiComponent component = draggingRight.get();
            double relativeX = cursor.guiX() - dragRightX;
            double relativeY = cursor.guiY() - dragRightY;
            dragRightX = cursor.guiX();
            dragRightY = cursor.guiY();
            component.gui().sendNewEvent(GuiEvent.DRAG_RIGHT,
                    new GuiComponentEvent(guiCursorX, guiCursorY, relativeX,
                            relativeY), component, engine);
        }
        double scrollX = controller.scrollX() * scrollSensitivity;
        double scrollY = controller.scrollY() * scrollSensitivity;
        if (scrollX != 0.0 || scrollY != 0.0) {
            engine.guiStack().fireRecursiveEvent(GuiEvent.SCROLL,
                    new GuiComponentEvent(guiCursorX, guiCursorY, scrollX,
                            scrollY, false), engine);
        }
        GuiComponentEvent componentEvent =
                new GuiComponentEvent(guiCursorX, guiCursorY);
        engine.guiStack()
                .fireEvent(componentEvent, GuiComponent::hover, engine);
        controller.pressEvents().forEach(event -> {
            switch (event.state()) {
                case PRESS:
                case REPEAT:
                    handlePress(event.key(), componentEvent);
                    break;
                case RELEASE:
                    handleRelease(event.key(), componentEvent);
                    break;
            }
        });
    }

    @Override
    public Stream<GuiCursor> cursors() {
        return Streams.of(cursor);
    }

    @Override
    public Stream<Pair<GuiCursor, ControllerBasic.PressEvent>> clicks() {
        return controller.pressEvents().map(event -> new Pair<>(cursor, event));
    }

    private void handlePress(ControllerKey key, GuiComponentEvent event) {
        switch (key) {
            case BUTTON_0:
                draggingLeft = engine.guiStack()
                        .fireEvent(GuiEvent.PRESS_LEFT, event, engine);
                dragLeftX = cursor.guiX();
                dragLeftY = cursor.guiY();
                if (engine.guiStack()
                        .fireEvent(GuiEvent.CLICK_LEFT, event, engine)
                        .isPresent()) {
                    return;
                }
                break;
            case BUTTON_1:
                draggingRight = engine.guiStack()
                        .fireEvent(GuiEvent.PRESS_RIGHT, event, engine);
                dragRightX = cursor.guiX();
                dragRightY = cursor.guiY();
                if (engine.guiStack()
                        .fireEvent(GuiEvent.CLICK_RIGHT, event, engine)
                        .isPresent()) {
                    return;
                }
                break;
            case KEY_ESCAPE:
                if (engine.guiStack().fireAction(GuiAction.BACK, engine)) {
                    return;
                }
                break;
            case KEY_ENTER:
                if (engine.guiStack().fireAction(GuiAction.ACTIVATE, engine)) {
                    return;
                }
                break;
            case KEY_UP:
                if (engine.guiStack().fireAction(GuiAction.UP, engine)) {
                    return;
                }
                break;
            case KEY_DOWN:
                if (engine.guiStack().fireAction(GuiAction.DOWN, engine)) {
                    return;
                }
                break;
            case KEY_LEFT:
                if (engine.guiStack().fireAction(GuiAction.LEFT, engine)) {
                    return;
                }
                break;
            case KEY_RIGHT:
                if (engine.guiStack().fireAction(GuiAction.RIGHT, engine)) {
                    return;
                }
                break;
        }
        firePress(key);
    }

    private void handleRelease(ControllerKey key, GuiComponentEvent event) {
        switch (key) {
            case BUTTON_0:
                if (draggingLeft.isPresent()) {
                    GuiComponent component = draggingLeft.get();
                    component.gui()
                            .sendNewEvent(GuiEvent.DROP_LEFT, event, component,
                                    engine);
                    draggingLeft = Optional.empty();
                }
                break;
            case BUTTON_1:
                if (draggingRight.isPresent()) {
                    GuiComponent component = draggingRight.get();
                    component.gui()
                            .sendNewEvent(GuiEvent.DROP_RIGHT, event, component,
                                    engine);
                    draggingRight = Optional.empty();
                }
                break;
        }
    }
}

File: src/main/kotlin/org/tobi29/scapes/engine/gui/GuiControllerMouse.kt
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.engine.gui

import org.tobi29.scapes.engine.ScapesEngine
import org.tobi29.scapes.engine.input.ControllerBasic
import org.tobi29.scapes.engine.input.ControllerDefault
import org.tobi29.scapes.engine.input.ControllerKey
import org.tobi29.scapes.engine.utils.math.vector.Vector2d

class GuiControllerMouse constructor(engine: ScapesEngine, controller: ControllerDefault,
                                     private val scrollSensitivity: Double = 1.0) : GuiControllerDefault(
        engine, controller) {
    private val cursor = GuiCursor()
    private var draggingLeft: GuiComponent? = null
    private var draggingRight: GuiComponent? = null
    private var dragLeftX = 0.0
    private var dragLeftY = 0.0
    private var dragRightX = 0.0
    private var dragRightY = 0.0

    override fun update(delta: Double) {
        val cursorX = controller.x()
        val cursorY = controller.y()
        cursor.set(Vector2d(cursorX, cursorY), Vector2d(cursorX, cursorY))
        draggingLeft?.let { component ->
            val guiPos = cursor.currentGuiPos()
            val relativeX = guiPos.x - dragLeftX
            val relativeY = guiPos.y - dragLeftY
            dragLeftX = guiPos.x
            dragLeftY = guiPos.y
            component.gui.sendNewEvent(GuiEvent.DRAG_LEFT,
                    GuiComponentEvent(cursorX, cursorY, relativeX,
                            relativeY), component)
        }
        draggingRight?.let { component ->
            val guiPos = cursor.currentGuiPos()
            val relativeX = guiPos.x - dragRightX
            val relativeY = guiPos.y - dragRightY
            dragRightX = guiPos.x
            dragRightY = guiPos.y
            component.gui.sendNewEvent(GuiEvent.DRAG_RIGHT,
                    GuiComponentEvent(cursorX, cursorY, relativeX,
                            relativeY), component)
        }
        val scrollX = controller.scrollX() * scrollSensitivity
        val scrollY = controller.scrollY() * scrollSensitivity
        if (scrollX != 0.0 || scrollY != 0.0) {
            engine.guiStack.fireRecursiveEvent(GuiEvent.SCROLL,
                    GuiComponentEvent(cursorX, cursorY, scrollX,
                            scrollY, false))
        }
        val componentEvent = GuiComponentEvent(cursorX, cursorY)
        engine.guiStack.fireEvent(componentEvent,
                GuiComponent::hover)
        controller.pressEvents().forEach { event ->
            when (event.state()) {
                ControllerBasic.PressState.PRESS, ControllerBasic.PressState.REPEAT -> handlePress(
                        event.key(), componentEvent)
                ControllerBasic.PressState.RELEASE -> handleRelease(event.key(),
                        componentEvent)
            }
        }
    }

    override fun cursors(): Sequence<GuiCursor> {
        return sequenceOf(cursor)
    }

    override fun clicks(): Sequence<Pair<GuiCursor, ControllerBasic.PressEvent>> {
        return controller.pressEvents().map { event -> Pair(cursor, event) }
    }

    private fun handlePress(key: ControllerKey,
                            event: GuiComponentEvent) {
        when (key) {
            ControllerKey.BUTTON_0 -> {
                val guiPos = cursor.currentGuiPos()
                draggingLeft = engine.guiStack.fireEvent(GuiEvent.PRESS_LEFT,
                        event)
                dragLeftX = guiPos.x
                dragLeftY = guiPos.y
                if (engine.guiStack.fireEvent(GuiEvent.CLICK_LEFT,
                        event) != null) {
                    return
                }
            }
            ControllerKey.BUTTON_1 -> {
                val guiPos = cursor.currentGuiPos()
                draggingRight = engine.guiStack.fireEvent(GuiEvent.PRESS_RIGHT,
                        event)
                dragRightX = guiPos.x
                dragRightY = guiPos.y
                if (engine.guiStack.fireEvent(GuiEvent.CLICK_RIGHT,
                        event) != null) {
                    return
                }
            }
            ControllerKey.KEY_ESCAPE -> if (engine.guiStack.fireAction(
                    GuiAction.BACK)) {
                return
            }
            ControllerKey.KEY_ENTER -> if (engine.guiStack.fireAction(
                    GuiAction.ACTIVATE)) {
                return
            }
            ControllerKey.KEY_UP -> if (engine.guiStack.fireAction(
                    GuiAction.UP)) {
                return
            }
            ControllerKey.KEY_DOWN -> if (engine.guiStack.fireAction(
                    GuiAction.DOWN)) {
                return
            }
            ControllerKey.KEY_LEFT -> if (engine.guiStack.fireAction(
                    GuiAction.LEFT)) {
                return
            }
            ControllerKey.KEY_RIGHT -> if (engine.guiStack.fireAction(
                    GuiAction.RIGHT)) {
                return
            }
        }
        firePress(key)
    }

    private fun handleRelease(key: ControllerKey,
                              event: GuiComponentEvent) {
        when (key) {
            ControllerKey.BUTTON_0 -> draggingLeft?.let { component ->
                component.gui.sendNewEvent(GuiEvent.DROP_LEFT, event, component)
                draggingLeft = null
            }
            ControllerKey.BUTTON_1 -> draggingRight?.let { component ->
                component.gui.sendNewEvent(GuiEvent.DROP_RIGHT, event,
                        component)
                draggingRight = null
            }
        }
    }
}
